### ISSUE TITLE
Imageset slicing is broken by batch offset

### ISSUE THREAD

**Issue by dagewa** [cctbx/dxtbx#633]
Discovered while investigating https://github.com/dials/dials/issues/2405 with @jmp1985 

Start with importing experiments that do not start from image 1
```
dials.import $(dials.data get -q insulin)/insulin_1_0{05..45}.img
```

Now load into python
```
>>> from dxtbx.model.experiment_list import ExperimentList
>>> el=ExperimentList.from_file("imported.expt")
>>> imageset=el.imagesets()[0]
```

The array range is (4, 45) as expected. Accessing imagesets is done with a zero-based index, as also expected
```
>>> imageset.get_array_range()
(4, 45)
>>> imageset[0]
(<scitbx_array_family_flex_ext.double object at 0x7fd8d06f6440>,)
```

But slicing does not work!
```
>>> imageset[0:1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/fcx32934/sw/cctbx/modules/dxtbx/src/dxtbx/imageset.py", line 330, in __getitem__
    return self.partial_set(reader, start, stop)
OverflowError: can't convert negative value to unsigned int
```

The first single image slice that works is 
```
>>> imageset[4:5]
<dxtbx_imageset_ext.ImageSequence object at 0x55e858f9ba30>
```

**Comment by graeme-winter** [comment]
?

**Comment by graeme-winter** [comment]
That is even more terse than some of your colleagues ;)

**Comment by dagewa** [comment]
:eyes:

**Comment by dagewa** [comment]
Oh it's a fun one. Just wait...

**Comment by graeme-winter** [comment]
This issue will _blow your mind_

**Comment by dagewa** [comment]
discovered while investigating https://github.com/dials/dials/issues/2405 with @jmp1985 

Start with importing experiments that do not start from image 1
```
dials.import $(dials.data get -q insulin)/insulin_1_0{05..45}.img
```

Now load into python
```
>>> from dxtbx.model.experiment_list import ExperimentList
>>> el=ExperimentList.from_file("imported.expt")
>>> imageset=el.imagesets()[0]
```

The array range is (4, 45) as expected. Accessing imagesets is done with a zero-based index, as also expected
```
>>> imageset.get_array_range()
(4, 45)
>>> imageset[0]
(<scitbx_array_family_flex_ext.double object at 0x7fd8d06f6440>,)
```

But slicing does not work!
```
>>> imageset[0:1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/fcx32934/sw/cctbx/modules/dxtbx/src/dxtbx/imageset.py", line 330, in __getitem__
    return self.partial_set(reader, start, stop)
OverflowError: can't convert negative value to unsigned int
```

The first single image slice that works is 
```
>>> imageset[4:5]
<dxtbx_imageset_ext.ImageSequence object at 0x55e858f9ba30>
```

**Comment by dagewa** [comment]
There are certainly similar things, but it seems like this was _introduced_ by the inclusion of batch offset handling. As @jmp1985 noted, this used to work

**Comment by dagewa** [comment]
I think the solution here is conceptually simple. Imagesets are _always_ accessed by a zero-based index, no matter what the image number or batch offset from the `Scan` is. That is also true of accessing like `imageset[0]`, or with `imageset.get_corrected_data(0)`. The `batch_offset`, by contrast, is a concept related to "image number", which is a series that might start from 1, but might also start from any other number.

I think the batch offset handling has no business being here, inside code related to `imageset` indexing and it should be removed.

**Comment by graeme-winter** [comment]
ISTR that the reason that this is there is the arrays were previously allocated from image[0,1] and bad things happened if your data were called image_202305040001.cbf (say)

i.e. we had (and have) a link between what the file is called and the array index. I humbly assert that _this_ is the underlying problem.

**Comment by dagewa** [comment]
Yes, the so-called `array_range` is _always_ equal to `(image_range[0] -1, image_range[1]`), but accessing the image data is done with respect to a zero-based array, no matter what the `array_range` says. 

I had gone through that in detail some time back (https://github.com/cctbx/dxtbx/issues/186#issuecomment-682071336) and decided that the naming is awful, but really the concept of index within the `array_range` is quite distinct from the actual accessing of data

**Comment by dagewa** [comment]
As discussed in DIALS-core, here's a branch to explore what happens when the imageset slicing is fixed: https://github.com/dials/dxtbx/tree/633-imageset-slicing

This leads to various test failures, as shown in the attached: 
[test-failures.zip](https://github.com/cctbx/dxtbx/files/11399971/test-failures.zip)

**Comment by dagewa** [comment]
The dxtbx error is easy to fix, as it was a test for erroneous behaviour that is redundant when imagesequence slicing is made sane. The dials errors look trickier...

**Comment by dagewa** [comment]
For reference, imageset slicing is also non-sane with stills. That's actually a good thing, because it indicates we don't have divergence in behaviour between scans and stills. They are both wrong :-)

```
dials.import $(dials.data get -q cunir_serial)/merlin0047_1700*.cbf
```
```
>>> from dxtbx.model.experiment_list import ExperimentList
>>> el=ExperimentList.from_file("imported.expt")
>>> imageset=el[0].imageset
>>> imageset[0]
(<scitbx_array_family_flex_ext.double object at 0x7f750355db40>,)
>>> imageset[1]
(<scitbx_array_family_flex_ext.double object at 0x7f750355d9c0>,)
>>> imageset[0:1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/fcx32934/sw/cctbx/modules/dxtbx/src/dxtbx/imageset.py", line 330, in __getitem__
    return self.partial_set(reader, start, stop)
OverflowError: can't convert negative value to unsigned int
```

**Comment by dagewa** [comment]
The DIALS work is going on in https://github.com/dials/dials/tree/remove-batch-offset-from-imageset-slicing. Down to 6 test failures now.

**Comment by dagewa** [comment]
This was fixed by #634
